### ISSUE TITLE
Test log to determine which test last ran when the test runner crashed

### ISSUE THREAD

**Issue by NightOwl888** [nunit/nunit3-vs-adapter#381]
### Problem

Occasionally, I have to try to figure out which of my nearly 8000 tests is causing the build runner to crash, which happens more frequently on .NET Core now because some exceptions cannot be caught. 

It is an extremely time consuming process to run smaller and smaller batches of tests to try to isolate which test or tests are causing the crash. It would be helpful if there were some kind of log so it would be possible to see which test (or tests) was running at the time of the crash. Right now the only solution seems to be to run smaller and smaller batches of tests until the one that causes the crash is isolated - which takes several hours to do when it takes an hour to run all of the tests, 30 minutes to run half of them, etc. Add to this that many of the tests are random, meaning they could crash on some runs and not others, and it is like finding a needle in a needlestack.

Or, if there is such a log somewhere (or easy way to build one), please excuse my ignorance and let me know where it is. Or if there is some way to determine which test was last running based on the fact that 219 of the 3428 tests passed before it crashed so the rest of them didn't run, let me know. Or if there is some way to determine what the next test that would run is based on the ones that passed, let me know - I know they always run in the same order, I just don't have any clue how to reverse engineer what the next test to run would be (but if I had a log, I would be able to look it up...).

Such a log could also potentially be used to determine if a test that failed on the current test run failed on the previous test run, etc.

### NUnit Version Info

NUnit Version: 3.7.1
NUnit3TestAdapter Version: 3.8.0
Microsoft.NET.Test.Sdk Version: 15.0.0

### Visual Studio Info

Microsoft Visual Studio Community 2017 
Version 15.3.2
VisualStudio.15.Release/15.3.2+26730.10
Microsoft .NET Framework
Version 4.7.02046

Installed Version: Community

Visual Basic 2017   00369-60000-00001-AA020
Microsoft Visual Basic 2017

Visual C# 2017   00369-60000-00001-AA020
Microsoft Visual C# 2017

Visual C++ 2017   00369-60000-00001-AA020
Microsoft Visual C++ 2017

Visual F# 4.1   00369-60000-00001-AA020
Microsoft Visual F# 4.1

.NET Reflector Visual Studio Extension   9.2.0.117
Integrates .NET Reflector into Visual Studio to allow you to seamlessly debug into third-party code and assemblies, even if you don't have the source code for them.

Visit www.reflector.net for more information.

Copyright (c) 2009-2017 Red Gate Software Inc.

Application Insights Tools for Visual Studio Package   8.8.00712.1
Application Insights Tools for Visual Studio

ASP.NET and Web Tools 2017   15.0.30726.0
ASP.NET and Web Tools 2017

ASP.NET Core Razor Language Services   1.0
Provides languages services for ASP.NET Core Razor.

ASP.NET Template Engine 2017   15.0.30726.0
ASP.NET Template Engine 2017

ASP.NET Web Frameworks and Tools 2017   5.2.50601.0
For additional information, visit https://www.asp.net/

Azure App Service Tools v3.0.0   15.0.30728.0
Azure App Service Tools v3.0.0

Azure Data Lake Node   1.0
This package contains the Data Lake integration nodes for Server Explorer.

Azure Data Lake Tools for Visual Studio   2.2.9000.1
Microsoft Azure Data Lake Tools for Visual Studio

Azure Data Lake Tools for Visual Studio   2.2.9000.1
Microsoft Azure Data Lake Tools for Visual Studio

Common Azure Tools   1.10
Provides common services for use by Azure Mobile Services and Microsoft Azure Tools.

JavaScript Language Service   2.0
JavaScript Language Service

JavaScript Project System   2.0
JavaScript Project System

Microsoft Azure HDInsight Azure Node   2.2.9000.1
HDInsight Node under Azure Node

Microsoft Azure Hive Query Language Service   2.2.9000.1
Language service for Hive query

Microsoft Azure Stream Analytics Language Service   2.2.9000.1
Language service for Azure Stream Analytics

Microsoft Azure Stream Analytics Node   1.0
Azure Stream Analytics Node under Azure Node

Microsoft Azure Tools   2.9
Microsoft Azure Tools for Microsoft Visual Studio 2017 - v2.9.50719.1

Microsoft Continuous Delivery Tools for Visual Studio   0.3
Simplifying the configuration of continuous build integration and continuous build delivery from within the Visual Studio IDE.

Microsoft JVM Debugger   1.0
Provides support for connecting the Visual Studio debugger to JDWP compatible Java Virtual Machines

Microsoft MI-Based Debugger   1.0
Provides support for connecting Visual Studio to MI compatible debuggers

Microsoft Visual C++ Wizards   1.0
Microsoft Visual C++ Wizards

Microsoft Visual Studio VC Package   1.0
Microsoft Visual Studio VC Package

Node.js Tools   1.0.0.0
Adds support for developing and debugging Node.js apps in Visual Studio

NuGet Package Manager   4.3.0
NuGet Package Manager in Visual Studio. For more information about NuGet, visit http://docs.nuget.org/.

SQL Server Data Tools   15.1.61707.200
Microsoft SQL Server Data Tools

ToolWindowHostedEditor   1.0
Hosting json editor into a tool window

TypeScript   2.3.4.0
TypeScript tools for Visual Studio

Visual Studio Code Debug Adapter Host Package   1.0
Interop layer for hosting Visual Studio Code debug adapters in Visual Studio

### .NET Platform

Any, but happens more frequently under .NET Core 1.0/2.0.

**Comment by CharliePoole** [comment]
This won't help you directly, since you are using the adapter, but it may give whomever works this issue some ideas... In the console runner, we can do that by specifying `--labels:before` so that the name of each test run displays as soon as it starts running.

The existing internal trace log can give quite a bit of information. In the console, it is activated by specifying `--trace:LEVEL`, where LEVEL would probably be Info or Verbose to get what you want. What I can't remember and have not yet checked is whether we hooked the facility up to the `.runsettings` file for the adapter. @OsirisTerje do you remember?

**Comment by OsirisTerje** [comment]
Have you tried to use the console version of the VS tester,  the vstest.console?  (If you're using the nuget adapter version, it will be picked up, if you're using the vsix, then just add the option /UseVsixExtension:true)
Using that one you will get a list of tests passed (or failed), and I would assume it will then just stop and crash when it hits the faulty test.

**Comment by NightOwl888** [comment]
I didn't know about vstest.console, but after locating it under my Visual Studio 2017 folder and trying to run the commands, it apparently doesn't support .NET Core/.NET Standard. 

```
Microsoft (R) Test Execution Command Line Tool Version 15.0.26720.2
Copyright (c) Microsoft Corporation.  All rights reserved.

/Framework:<Framework Version>
      Target .Net Framework version to be used for test execution.
      Valid values are Framework35, Framework40 and Framework45
```

I just ran the tests under .NET Framework 4.5.1 in Visual Studio 2017 and in this instance they run all the way through (although, many of the tests failed with OutOfMemoryException, which is another issue I have running in Visual Studio that doesn't happen on the command line - perhaps there is a setting to increase the heap size somewhere...?).

I tried running the .NET Core 1.0 tests using dotnet test, but in that environment they run all the way through without crashing.  But they still crash in Visual Studio 2017 and that is what needs to be fixed.

I have isolated 1 of the problems. `Debug.Assert(false)` causes the build runner to crash on .NET Standard 2.0 with a .NET Core 2.0 client. I bypassed reporting here and [reported directly to the VSTest team](https://github.com/Microsoft/vstest/issues/1022) because the problem also happens when using xUnit. However, I tried changing to Release mode (removing all of the Debug.Asserts) and running the tests and they are still crashing with another unknown cause.

There were issues when using the .NET SDK 1.0.0-preview2-1-003177 tooling, too, but I decided to postpone working on them until switching to the new `.csproj` format (which I am in the middle of). And it seems that upgrading didn't automatically fix those crashes as I had hoped.

Unfortunately, I haven't found a way to make these crashes happen anywhere but Visual Studio, so ideally I could somehow output a log. If there are no switches to turn on logging anywhere, are there at least some hooks so I can add it? Or is my only option to add this feature to a base class of all of the tests?

**Comment by OsirisTerje** [comment]
Have you tried to enable the tracing feature in VSTest?  
Open the config files for vstest.executionengine.exe/vstest.discoveryengine.exe, there is a setting there for tracing/verbosity, just set that to 4.

**Comment by OsirisTerje** [comment]
You could also enable the dump feature, it will write out a dump file per assembly tested.  It only do this for .net framework though, not .net core. See DumpXmlTestResults   in https://github.com/nunit/docs/wiki/Tips-And-Tricks

**Comment by OsirisTerje** [comment]
Done : Means Use the dump feature in 3.9.0
Won't fix : Means no further fix planned.

**Comment by sanderaernouts** [comment]
_I know this is quite an old issue but googling led me here while investigating a similar issues with tests resulting in a `StackOverflowException` on the CI server._

To figure out which test was failing I used `vstest.console.exe` with the `/Blame` option (see [this link](https://docs.microsoft.com/en-us/visualstudio/test/vstest-console-options?view=vs-2019)). I did have to set the `NumberOfTestWorkers` to `1` to narrow down the actual test cases that were causing the crash. I use parallel execution in NUnit which resulted in `vstest.console.exe` blaming one of the tests that was running in parallel at the time of the crash but that was not always the one that actually caused the error.

I used these run settings to restrict the number of parallel executions:
```
<?xml version="1.0" encoding="utf-8"?>
<RunSettings>
  <NUnit>
    <NumberOfTestWorkers>1</NumberOfTestWorkers>
  </NUnit>
</RunSettings>
```
I used these steps to figure out which test was the problem:
1. run `vstest.console.exe` with the `/Blame` option against all test assemblies to figure out which assembly contains the test or tests that are crashing
2. run `vstest.console.exe` with the `/Blame` option against that specific assembly with a specific run settings file that restricts the parallel test execution in NUnit to 1.

Hope this helps the next person trying to figure out which test is crashing :)!
